### PR TITLE
[SPARK-25548][SQL]In the PruneFileSourcePartitions optimizer, replace the nonPartitionOps field with true in the And(partitionOps, nonPartitionOps) to make the partition can be pruned

### PR DESCRIPTION
**Pull request by eatoncys**
## What changes were proposed in this pull request?
In the PruneFileSourcePartitions optimizer, the partition files will not be pruned if we use partition filter and non partition filter together, for example:

    sql("CREATE TABLE IF NOT EXISTS src_par (key INT, value STRING) partitioned by(p_d int) stored as parquet ")
    sql("insert overwrite table src_par partition(p_d=2) select 2 as key, '4' as value")
    sql("insert overwrite table src_par partition(p_d=3) select 3 as key, '4' as value")
    sql("insert overwrite table src_par partition(p_d=4) select 4 as key, '4' as value")

Before this PR, the sql below will scan all the partition files, in which, the partition **p_d=4** should be pruned.

    sql("select * from src_par where (p_d=2 and key=2) or (p_d=3 and key=3)").show

After this PR, the partition **p_d=4** will be pruned

## How was this patch tested?
exist test

### CODE REVIEW

**Comment by eatoncys** [comment]
cc @gatorsmile @cloud-fan

**Comment by dilipbiswal** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala [hunk 1]
```diff
@@ -39,21 +40,31 @@ private[sql] object PruneFileSourcePartitions extends Rule[LogicalPlan] {
...
       // case insensitive, we should change them to match the one in schema, so we donot need to
       // worry about case sensitivity anymore.
       val normalizedFilters = filters.map { e =>
-        e transform {
+        e transformUp {
           case a: AttributeReference =>
             a.withName(logicalRelation.output.find(_.semanticEquals(a)).get.name)
+          // Replace the nonPartitionOps field with true in the And(partitionOps, nonPartitionOps)
+          // to make the partition can be pruned
+          case and @And(left, right) =>
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala]: nit: and @ And

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala [hunk 2]
```diff
@@ -39,21 +40,31 @@ private[sql] object PruneFileSourcePartitions extends Rule[LogicalPlan] {
...
+            val leftPartition = left.references.filter(partitionSet.contains(_))
+            val rightPartition = right.references.filter(partitionSet.contains(_))
+            if (leftPartition.size == left.references.size && rightPartition.size == 0) {
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala]: `rightPartition.size == 0` => `rightPartition.isEmpty`

**Review by xuanyuanking — COMMENTED** [review]
Is it possible to add a UT in PruneFileSourcePartitionsSuite to check the optimized rule take effect?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala [hunk 3]
```diff
@@ -39,21 +40,31 @@ private[sql] object PruneFileSourcePartitions extends Rule[LogicalPlan] {
...
+            } else and
         }
       }
-
-      val sparkSession = fsRelation.sparkSession
-      val partitionColumns =
-        logicalRelation.resolve(
-          partitionSchema, sparkSession.sessionState.analyzer.resolver)
-      val partitionSet = AttributeSet(partitionColumns)
       val partitionKeyFilters =
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala]: I think a simpler change is
```
val partitionKeyFilters = ExpressionSet(normalizedFilters.map(splitConjunctivePredicates).filter...)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala (hunk 2 above)
Comment (eatoncys) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala]: Ok, thanks for review.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala (hunk 3 above)
Comment (eatoncys) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala]: @cloud-fan Sorry, I don't understand very clearly, the function of splitConjunctivePredicates can only split and(a,b); if there is a or expression in the filter , for example 'where (p_d=2 and key=2) or (p_d=3 and key=3)', the result of splitConjunctivePredicates is '(((p_d#2 = 2) && (key#0 = 2)) || ((p_d#2 = 3) && (key#0 = 3)))', the partition expression could not be split out.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala (hunk 2 above)
Comment (eatoncys) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala]: Ok, I will add a UT latter, thanks.

**Comment by cloud-fan** [comment]
Please describe the idea clearly in the PR description.

After reading the code, I think the idea is, for any predicate `p`, we can rewrite it to `p'` that only contains partition columns. The requirement is, partitions returned by the partition predicate `p`, must also be returned by `p'`.

We need to be careful here, to not introduce correctness issues. Can you prove that your rewrite is safe?

**Comment by eatoncys** [comment]
@cloud-fan which should be proved is that: the partitions returned of p' shound contain the partitions returned by p.  Here, let p' = p && x, if x is true then p' == p; else if x is false, the paritions returned we needed is none, so p' contains the partitions we needed.

**Comment by cloud-fan** [comment]
Your rewrite is applied to the entire predicate expression tree, how about the case `NOT(p AND x)`? If `p` is true, `x` is false, then `NOT(p AND x)` is true. But `NOT(true AND true)` is false.

**Comment by eatoncys** [comment]
@cloud-fan Yes, it has problems for Not expression, we need find some good ways. Thanks for review.

**Pull request closed without merging**